src/frontend/src/optimizer/property/func_dep.rs

```diff
@@ -0,0 +1,338 @@
+    ) {
+        let from = {
+            let mut tmp = FixedBitSet::with_capacity(column_cnt);
+            for &i in from {
+                tmp.set(i, true);
+            }
+            tmp
+        };
+        let to = {
+            let mut tmp = FixedBitSet::with_capacity(column_cnt);
+            for &i in to {
+                tmp.set(i, true);
+            }
+            tmp
+        };
+        self.add_functional_dependency(FunctionalDependency::new(from, to))
+    }
+
+    fn get_closure(&self, columns: FixedBitSet) -> FixedBitSet {
+        let mut closure = columns;
+        let mut no_updates;
+        loop {
+            no_updates = true;
+            for FunctionalDependency { from, to } in &self.strict {
+                if from.is_subset(&closure) && !to.is_subset(&closure) {
+                    closure.union_with(to);
+                    no_updates = false;
+                }
+            }
+            if no_updates {
+                break;
+            }
+        }
+        closure
+    }
+
+    /// Return `true` if the dependency determinant -> dependant exists.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use risingwave_frontend::optimizer::property::FunctionalDependencySet;
+    /// # use fixedbitset::FixedBitSet;
+    /// let mut fd = FunctionalDependencySet::new();
+    /// fd.add_functional_dependency_by_column_indices(&[1, 2], &[0], 5); // (1, 2) --> (0)
+    /// fd.add_functional_dependency_by_column_indices(&[0, 1], &[3], 5); // (0, 1) --> (3)
+    /// fd.add_functional_dependency_by_column_indices(&[3], &[4], 5); // (3) --> (4)
+    /// let from = FixedBitSet::from_iter([1usize, 2usize].into_iter());
+    /// let to = FixedBitSet::from_iter([4usize].into_iter());
+    /// assert!(fd.is_determined_by(from, to)); // (1, 2) --> (4) holds
+    /// ```
+    pub fn is_determined_by(&self, determinant: FixedBitSet, dependant: FixedBitSet) -> bool {
+        self.get_closure(determinant).is_superset(&dependant)
+    }
+
+    /// Return true if `columns` is a key.
+    pub fn is_key(&self, columns: FixedBitSet) -> bool {
+        let all_columns = {
+            let mut tmp = columns.clone();
+            tmp.set_range(.., true);
+            tmp
+        };
+        self.is_determined_by(columns, all_columns)
+    }
+
+    /// Remove redundant columns from the given set.
+    /// Redundant columns can be functionally determined by other columns so there is no need to
+    /// keep them in a key. Note that the accurate minimization algorithm can take O(2^n) time,
+    /// so we use a approximate algorithm.
+    pub fn minimize_key(&self, key: FixedBitSet) -> FixedBitSet {
+        let mut new_key = key.clone();
+        for i in key.ones() {
+            new_key.set(i, false);
+            if !self.is_key(new_key.clone()) {
+                new_key.set(i, true);
+            }
+        }
+        new_key
+    }
+
+    pub fn rewrite_with_mapping(mut self, col_change: ColIndexMapping) -> Self {
```

Comment:
maybe it is better to implement this method on `ColIndexMapping` like Order or Distribution